app/experimenter/nimbus-ui/src/components/PageEditBranches/FormBranches/reducer/actions.ts

```diff
@@ -250,17 +257,117 @@ function commitFormData(
 
   if (referenceBranch) {
+    referenceBranch = branchUpdatedWithFormData(
+      referenceBranch,
+      formData.referenceBranch,
+    );
+  }
+
+  if (treatmentBranches) {
+    treatmentBranches = treatmentBranches.map((treatmentBranch, idx) =>
+      branchUpdatedWithFormData(
+        treatmentBranch,
+        formData.treatmentBranches?.[idx],
+      ),
+    );
+  }
+
+  return {
+    ...state,
+    referenceBranch,
+    treatmentBranches,
+  };
+}
+
+function branchUpdatedWithFormData(
+  branch: AnnotatedBranch,
+  formData: Partial<AnnotatedBranch> | null | undefined,
+) {
+  const screenshots = branch.screenshots?.map((screenshot, idx) => {
+    const { description, image } = formData?.screenshots?.[idx] || {};
+    return {
+      id: screenshot?.id,
+      description,
+      image,
+    };
+  });
+  return {
+    ...branch,
+    ...formData,
+    screenshots,
+  };
+}
+
+type AddScreenshotToBranchAction = {
+  type: "addScreenshotToBranch";
+  branchIdx: number;
+};
+
+function addScreenshotToBranch(
+  state: FormBranchesState,
+  action: AddScreenshotToBranchAction,
+) {
+  const { branchIdx } = action;
+  let { referenceBranch, treatmentBranches } = state;
+
+  const newScreenshot: BranchScreenshotType = { description: "", image: null };
+
+  if (branchIdx === REFERENCE_BRANCH_IDX && referenceBranch) {
     referenceBranch = {
       ...referenceBranch,
-      ...formData.referenceBranch,
+      screenshots: [...(referenceBranch.screenshots || []), newScreenshot],
     };
+  } else if (treatmentBranches) {
+    treatmentBranches = withModifiedBranch(
+      treatmentBranches,
+      branchIdx,
+      (selectedBranch) => ({
+        ...selectedBranch,
+        screenshots: [...(selectedBranch.screenshots || []), newScreenshot],
+      }),
+    );
   }
 
-  if (treatmentBranches) {
-    treatmentBranches = treatmentBranches.map((treatmentBranch, idx) => ({
-      ...treatmentBranch,
-      ...formData.treatmentBranches?.[idx],
-    }));
+  return {
+    ...state,
+    referenceBranch,
+    treatmentBranches,
+  };
+}
+
+type RemoveScreenshotFromBranchAction = {
+  type: "removeScreenshotFromBranch";
+  branchIdx: number;
+  screenshotIdx: number;
+};
+
+function removeScreenshotFromBranch(
+  state: FormBranchesState,
+  action: RemoveScreenshotFromBranchAction,
+) {
+  const { branchIdx, screenshotIdx } = action;
+  let { referenceBranch, treatmentBranches } = state;
+
+  if (branchIdx === REFERENCE_BRANCH_IDX && referenceBranch) {
+    referenceBranch = {
+      ...referenceBranch,
+      screenshots: withoutScreenshot(
+        referenceBranch.screenshots,
+        screenshotIdx,
+      ),
+    };
+  } else if (treatmentBranches) {
+    treatmentBranches = withModifiedBranch(
+      treatmentBranches,
+      branchIdx,
+      (selectedBranch) => ({
+        ...selectedBranch,
+        screenshots: withoutScreenshot(
+          selectedBranch.screenshots,
+          screenshotIdx,
+        ),
+      }),
+    );
```

Comment:
It would be nice if there was some way to treat all branches holistically for cases like this where you don't really care if it's the reference or treatment branches.  Nothing we should address here but maybe there's a pattern we can find later so we don't always have to treat reference as a special case?